bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitHeaterCoolerImpl.java

```diff
@@ -0,0 +1,193 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.io.homekit.internal.accessories;
+
+import static org.openhab.io.homekit.internal.HomekitCharacteristicType.CURRENT_HEATER_COOLER_STATE;
+import static org.openhab.io.homekit.internal.HomekitCharacteristicType.TARGET_HEATER_COOLER_STATE;
+
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.items.GenericItem;
+import org.eclipse.smarthome.core.library.items.StringItem;
+import org.eclipse.smarthome.core.library.items.SwitchItem;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.OpenClosedType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.io.homekit.internal.HomekitAccessoryUpdater;
+import org.openhab.io.homekit.internal.HomekitCharacteristicType;
+import org.openhab.io.homekit.internal.HomekitSettings;
+import org.openhab.io.homekit.internal.HomekitTaggedItem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.github.hapjava.accessories.HeaterCoolerAccessory;
+import io.github.hapjava.characteristics.HomekitCharacteristicChangeCallback;
+import io.github.hapjava.characteristics.impl.heatercooler.CurrentHeaterCoolerStateEnum;
+import io.github.hapjava.characteristics.impl.heatercooler.TargetHeaterCoolerStateEnum;
+import io.github.hapjava.characteristics.impl.thermostat.TemperatureDisplayUnitCharacteristic;
+import io.github.hapjava.characteristics.impl.thermostat.TemperatureDisplayUnitEnum;
+import io.github.hapjava.services.impl.HeaterCoolerService;
+
+/**
+ * Implements Heater Cooler
+ *
+ * @author Eugen Freiter - Initial contribution
+ */
+
+public class HomekitHeaterCoolerImpl extends AbstractHomekitAccessoryImpl implements HeaterCoolerAccessory {
+    private final Logger logger = LoggerFactory.getLogger(HomekitHeaterCoolerImpl.class);
+    private final BooleanItemReader activeReader;
+    private final EnumMap<CurrentHeaterCoolerStateEnum, String> currentStateMapping = new EnumMap(
```

Comment:
```suggestion
    private final Map<CurrentHeaterCoolerStateEnum, String> currentStateMapping = new EnumMap(
```

Reply:
i will change to map. but just for my understanding / learning - what is the advantage of using Map here?  compiler will probably replace Map with EnumMap at build time or?

Reply:
Generally you want your code to define behavior without being beholden to any particular concrete implementation. That is why you should interact with `List`, `Set`, `Map`, or `Collection` instead of `ArrayList`, `LinkedList`, `HashMap`, etc...

Interacting with interfaces does a better job of displaying the code's intent without getting bogged down with the concrete implementations of your data structures.

Reply:
ok. understood. no optimisation but improving readability and best practice.